components-java-sdk/src/main/java/io/dapr/components/domain/pubsub/PullMessagesRequest.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2023 The Dapr Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.dapr.components.domain.pubsub;
+
+import java.util.Optional;
+
+public record PullMessagesRequest(Topic topic, String ackMessageId, Optional<String> ackErrorMessage) {
```

Comment:
One pattern that I took while developing the StateStore domain classes and also took here is that if a given structure holds a single field (like `AckMessageError`, that holds a single `message`) I just bypass the indirection and embed the fields directly here.

https://github.com/dapr/dapr/blob/master/dapr/proto/components/v1/pubsub.proto#L46-L49

Also did something similar to empty messages (like `PingResponse`, `PublishResponse` and such): just don't model them at all and map them to `Mono<Void>`.

Would that cause problems later down the road when we potentially extend those types? Potentially, as we would need to break the API here to add these types. But right now, overcomplicating to keep a 1:1 mapping to the .proto definitions seems like an overkill and bad DevEx. Thoughts?

Reply:
On a separate note: the protocol buffer this object represents behaves as a union: its first use requires a given field to be present but subsequent messages require a disjoint subset of fields to be set.

Instead of jumping into "how to represent  C-like unions in Java", I thought about splitting these these into 2 disjoint types that have to be provided as parameters int the `pullMessage` fields. The PubSub interface would be modified to have this method

```java
    Flux<PullMessagesResponse> pullMessages(Topic requestedTopic, Flux<PullMessagesAcknowledgement> acks);
```

@philliphoff  and @mcandeia, thoughts?

Reply:
The .NET SDK does something similar, with the adaptor handling receipt of the initial topic request, so the actual pluggable component need not do it.  Then, the component can use a callback (the "delivery handler") to push messages down to the caller, as well as pass an additional callback handles acknowlegements (either getting passed an error message or not).  While I've tried not go overboard in other places by "flattening" the `.proto` requests/responses, this is one place where the SDK can make the developer experience much nicer.

```csharp
Task PullMessagesAsync(PubSubPullMessagesTopic topic, MessageDeliveryHandler<string?, PubSubPullMessagesResponse> deliveryHandler, CancellationToken cancellationToken = default);
```

Reply:
> Would that cause problems later down the road when we potentially extend those types? Probably, as we would need to break the API to add these types. But right now, overcomplicating to keep a 1:1 mapping to the .proto definitions seems like an overkill and bad DevEx. Thoughts?

I think there are definitely places where "flattening" or "unwrapping" the requests/responses makes sense, and the .NET SDK does similar things in some places (e.g. error messages and ETags values).  I think we just have to weight the likelihood of any given request/response changing over time, and whether such changes could also be managed in different ways (e.g. exposed as additional "V2" and "V3" interfaces).